src/lib/components/grid.tsx

```diff
@@ -89,8 +97,19 @@ export class ReactGrid extends Component<IGridProps, IGridState> {
   componentDidMount() {
     this.ensureGridContainerInDOM();
     this.grid.build(this.gridContainer);
-    this.reflectNewRowsOrCols(this.props.rows, this.grid.rows);
-    this.reflectNewRowsOrCols(this.props.cols, this.grid.cols);
+    this.cellRendererBuilder = this.grid.colModel.createBuilder(
+      () => document.createElement('div'),
+      (element, { data, virtualCol, virtualRow }) => {
+        const rendered = this.props.cellRenderer && this.props.cellRenderer(virtualRow, virtualCol, data);
```

Comment:
Could extraneous `createElement` calls be avoided by short-circuiting if `cellRenderer` was called first and returned `null`, or `cellRenderer` is undefined?

Reply:
if so that could be interesting but it would mean we'd be treating the first call to cellRenderer for each row, col as the source of truth. so if you didn't render something the first call but then conditionally wanted to render something later it would get called.

Reply:
(oh)

Reply:
i'm curious what your idea was tho cause i may be misunderstanding

Reply:
I thought you mentioned in #1 that the the containing div doesn't need to be created unless something is going to operate on it (else it's a cheaper text node). So the idea would be to short-circuit that call if we don't need the `div`. The only way I see how to do that in retrospect is to make the first function also aware of the virtualRow/col positions in order to preemptively call render, which I am not sure is possible in the current way of doing things.

Reply:
oh ya the containing div could be avoided but only if that cell would never be used to render anything but text. that's why the builders are designed to be per column. the idea is that mostly that type of content would apply to an entire column (more one off-y stuffy can often be done with style rather than html to help with perf). the per column design was intended to encourage that pattern. that said, if the renderer returns text or undefined the grid will ultimately fall back to a text node. the vast majority of the slowness we're seeing here is coming from rendering the react components into the container element not the initial container create. so as long as the cell renderer doesn't do that work the grid will still optimize appropriately if that makes sense.